README.md

```diff
@@ -1,16 +1,1605 @@
+  # inferred from the invocation task audience.
+  iss     optional Principal
+
+  # When issuer is different from executor this MUST hold a UCAN
+  # delegation chain from executor to the issuer. Should be omitted executor is an issuer.
+  prf     [&UCAN] implicit ([])
+
+  # Signature from the `iss`.
+  s       Varsig
+}
+
+type Result<Ok, Error> union {
+  | Ok ("ok") # Success
+  | Error ("error") # Error
+} representation kinded
+
+type Promise<In> union {
+  | &Task<In>          ("ucan/task")
+  | &Invocation<In>    ("ucan/invocation")
+} representation keyed
+
+# Promise is a way to reference invocation receipt
+type Await<In> union {
+  &Promise<In>    "await/*"
+  &Promise<In>    "await/ok"
+  &Promise<In>    "await/error"
+} representation keyed
+```
+
+# 3 Task
+
+A Task is the smallest unit of work that can be requested from a UCAN. It describes one `(resource, ability, input)` triple. The `input` field is free form, and depend on the specific resource and ability being interacted with, and is not described in this specification.
+
+Using the JavaScript analogy from the introduction, a Task is similar to wrapping a call in an anonymous function:
+
+```json
+{
+  "v": "0.1.0",
+  "with": "mailto://alice@example.com",
+  "do": "msg/send",
+  "input": {
+    "to": ["bob@example.com", "carol@example.com"],
+    "subject": "hello",
+    "body": "world"
+  },
+  "prf": [
+    { "/": "bafyreibblnq5bawcchzh73nxkdmkx47hu64uwistvg4kyvdgfd6igkcnha" }
+  ]
+}
+```
+
+```js
+// Pseudocode
+() => msg.send("mailto:alice@example.com", {
+  to: ["bob@example.com", "carol@example.com"],
+  subject: "hello",
+  body: "world"
+})
+```
+
+Later, when we explore [Promise]s, this also includes capturing the promise:
+
+```json
+{
+  "bafyreihroupaenuxjduzs4ynympv3uawcct4mj4sa7ciuqqlss4httehiu": {
+    "v": "0.1.0",
+    "do": "crud/read",
+    "with": "https://example.com/mailinglist",
+    "prf": [
+      { "/": "bafyreib2dldcev74g5i76bacd4w72gowuraouv2kvnwa2ympvrvtybcsfi" }
+    ]
+  },
+  "bafyreigiz22kfo4bbrsl3jyspm5ykuh2jk5hxmduc5yzijp3dzegvvi6tq": {
+    "v": "0.1.0",
+    "with": "mailto://alice@example.com",
+    "do": "msg/send",
+    "input": {
+      "to": {
+        "await/*": {
+          "ucan/task": {
+            "/": "bafyreihroupaenuxjduzs4ynympv3uawcct4mj4sa7ciuqqlss4httehiu"
+          }
+        }
+      },
+      "subject": "hello",
+      "body": "world"
+    },
+    "prf": [
+      { "/": "bafyreibblnq5bawcchzh73nxkdmkx47hu64uwistvg4kyvdgfd6igkcnha" }
+    ]
+  }
+}
+```
+
+```js
+// Pseudocode
+const mailingList = crud.read("https://exmaple.com/mailinglist")
+msg.send("mailto:alice@example.com", {
+  to: (await mailingList).ok,
+  subject: "hello",
+  body: "world",
+})
+```
+
+## 3.1 Schema
+
+```ipldsch
+type Task<In> struct {
+  v       SemVer
+
+  with    URI
+  do      Ability
+
+  input   In (implicit {})
+  meta    {String : Any} (implicit {})
+  nnc     string (implicit "")
+
+  prf     [&UCAN] (implicit [])
```

Comment:
🤔 Since there's no `iss` in here anymore, should this be moved to `Invocation`?

Reply:
I thought issuer is derived from the `UCAN` itself. I'm amendable to moving it to `Invocation` but that would make things bit more complicated IMO because `Invocation` will become tuple of `(task, proofs, auth)` as opposed to just being `(task, auth)`. That in turn would make deriving Receipt addressed and stuff more complicated

Reply:
That makes sense. Perhaps some context: at least IPVM and Bacalhau (and possibly also WarpForge and Koii) need a stable identifier for Tasks independent of who ran them. This lets us look up all runs of that Task previously in a DHT, even by just saying "these arguments were applied to this function".

This stable identifier needs to include the following, and nothing else (not even a version):

```haskell
type Minimal<Args> struct {
  with    URI
  do      Ability
  input   Args (implicit {})
  nnc     string (implicit "")
}
```

As a motivating example: I have a Wasm module that I need to run, and I want to skip execution if it was previously executed by anyone else on the network. I don't know if it was run before, but when I plug it into Bacalhau's Wasm runtime, I see that it was run before, and I get a bunch of receipts that all concur on the final value.

We can define this separately, but it would be nice to be able to share a DHT for receipt lookup since everything you need to identify a task uniquely is in this minimal struct's CID.

Reply:
Sounds like removing `prf` is not going to be enough anyway as there still `meta` and v`. Good news is deriving DHT key from the `Task` would be straight forward. That said I'm not sure we'll end up keying receipts by such an ID, we're more likely to key them by `&Invocation` instead.

Can we pun on this and perhaps revisit in new versions after we had chance of assessing running implementations ?